### ISSUE TITLE
Formula list freezing on latest update of Excel 2016

### ISSUE THREAD

**Issue by mfriedy** [Excel-DNA/IntelliSense#42]
In the last few weeks, I have noticed that the Excel is now hanging for ~15secs if I have an addin loaded with intellisense added. In general, it occurs on the 2nd or 3rd character - in this example i was trying to type "NORM.DIST" and it hung when I typed "R". I am on Excel 7571.2058-64bit

I ran the code through visual studio to see if I could work out what is going wrong... and I think its occuring at this point: "PopupListWatcher.cs"

`listBounds = (Rect)_popupList.GetCurrentPropertyValue(AutomationElement.BoundingRectangleProperty);
Debug.Print($"#### PopupList Update - ListBounds: {listBounds} / SelectedItemBounds:{selectedItemBounds}");`

And the output log:
@>@>@>@> LoadNotification: Loaded - C:\Program Files\Microsoft Office\root\Office16\MSOHEV.DLL
STATE UPDATE (FormulaEditStart): 
			Ready 
		=>	FormulaEdit
	FormulaEditWindow: 658722
	FormulaPrefix: =
	EditWindowBounds: 485,193,2075,66
	ExcelToolTipWindow: 0
Created ToolTipForm with owner 658722
IntelliSenseDisplay - FormulaEditStart - FormulaEditWindow: 658722, ArgumentsToolTip: ExcelDna.IntelliSense.ToolTipForm, Text: 
^^^ FormulaEditStateChanged. CurrentPrefix: =, Thread 1
Not processing formula =n
POPUPLISTWATCHER WINDOW CURRENT SELECTION System.Windows.Automation.AutomationElement
#### PopupList Update - ListBounds: 504,262,159,206 / SelectedItemBounds: 505,263,140,17
STATE UPDATE (FormulaEditExcelToolTipChange): 
			FormulaEdit
	FormulaEditWindow: 658722
	FormulaPrefix: =n
	EditWindowBounds: 485,193,2075,66
	ExcelToolTipWindow: 0 
		=>	FormulaEdit
	FormulaEditWindow: 658722
	FormulaPrefix: =n
	EditWindowBounds: 485,193,2075,66
	ExcelToolTipWindow: 21105186
STATE UPDATE (FunctionListShow): 
			FormulaEdit
	FormulaEditWindow: 658722
	FormulaPrefix: =n
	EditWindowBounds: 485,193,2075,66
	ExcelToolTipWindow: 21105186 
		=>	FunctionList
	FunctionListWindow: 1444754
	FunctionListBounds: 504,262,159,206
	SelectedItemText: N
	SelectedItemBounds: 505,263,140,17
	FormulaEditWindow: 658722
	FormulaPrefix: =n
	EditWindowBounds: 485,193,2075,66
	ExcelToolTipWindow: 21105186
IntelliSenseDisplay - FunctionListShow
Created ToolTipForm with owner 1444754
Not processing formula =no
POPUPLISTWATCHER WINDOW CURRENT SELECTION System.Windows.Automation.AutomationElement

**Comment by govert** [comment]
Thank you for reporting this.
This kind of thing may be hard to track down, but you can temporarily disable IntelliSense by setting an environment variable

  `EXCELDNA_INTELLISENSE_DISABLEDVERSIONS=*`

Are you sure you are running the current version (1.0.0) ?
What Windows version are you on?

Was it working fine before - was there a time when you had the IntelliSense but not the delays?
How often does this happen?

I'll have to check more carefully, but the code you show should probably not be running on the Excel thread - so I'm not sure why it would hang Excel.

I think it's possible to enable a verbose logging mode, and that might give us some timing info too.

**Comment by mfriedy** [comment]
Yes, 1.0.0 - I downloaded the latest version of the code before trying it in debug mode. Am on W10, 10.0.14393 (i.e. not insider). Yes, this was definitely working at some point before, I think it was in the last 2-3 weeks things stopped working and now happens 100% of the time when I use the addin on my machine.

More than happy to run more verbose debug tests if it will help (if you can indicate how)! My guess about the point in the code comes from just pressing the pause button on the debugger (and also checking later with breakpoints around it)

**Comment by govert** [comment]
To add verbose logging, create a file called `ExcelDna.IntelliSense.xll.config` next to your `ExcelDna.IntelliSense.xll`. Put this into the config file:

```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <system.diagnostics>
    <trace autoflush="true" indentsize="4"/>
    <sources>
      <source name="ExcelDna.IntelliSense" switchValue="Verbose">
        <listeners>
          <remove name="Default" />
          <add name="File" type="System.Diagnostics.TextWriterTraceListener" initializeData="ExcelDna.IntelliSense.log" traceOutputOptions = "DateTime" >
            <filter type="System.Diagnostics.EventTypeFilter" initializeData="All"/>
          </add>
        </listeners>
      </source>
    </sources>
  </system.diagnostics>
</configuration>
```

You should see a log file called `ExcelDna.IntelliSense.log`. It might get big if you leave the logging on.

**Comment by mfriedy** [comment]
Hi Govert, 
I have attached a copy of the log file: [ExcelDna.IntelliSense.log.txt](https://github.com/Excel-DNA/IntelliSense/files/636862/ExcelDna.IntelliSense.log.txt)

However, i think the key section is here:

~~~~ExcelDna.IntelliSense Verbose: 4 : >>>> FormulaEditWatcher.UpdateFormula CurrentPrefix: =su
    DateTime=2016-12-07T14:42:48.8811874Z
ExcelDna.IntelliSense Verbose: 4 : FormulaEdit UpdateEditState Formula editing: CurrentPrefix =su, EditWindowBounds: 485,193,2075,66
    DateTime=2016-12-07T14:42:48.9030406Z
ExcelDna.IntelliSense Verbose: 2 : !> ExcelToolTip ToolTipChanged Received: Show:0x659458 with state: FormulaEdit
    DateTime=2016-12-07T14:42:48.9030406Z
ExcelDna.IntelliSense Verbose: 4 : PopupList window show
    DateTime=2016-12-07T14:42:48.9040437Z
ExcelDna.IntelliSense Verbose: 4 : PopupList PopupListElementSelectedHandler on thread 7
    DateTime=2016-12-07T14:42:48.9394791Z
ExcelDna.IntelliSense Verbose: 4 : PopupList PopupListElementSelectedHandler on thread 1
    DateTime=2016-12-07T14:42:48.9454690Z
ExcelDna.IntelliSense Warning: 4 : PopupList - Could not update selected item: System.Runtime.InteropServices.COMException (0x80131505): Operation timed out. (Exception from HRESULT: 0x80131505)
   at UIAutomationClient.IUIAutomationElement.GetCurrentPropertyValueEx(Int32 propertyId, Int32 ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property, Boolean ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property)
   at ExcelDna.IntelliSense.PopupListWatcher.UpdateSelectedItem(AutomationElement newSelectedItem)
    DateTime=2016-12-07T14:43:08.9484830Z
ExcelDna.IntelliSense Warning: 4 : PopupList - Could not update selected item: System.Runtime.InteropServices.COMException (0x80131505): Operation timed out. (Exception from HRESULT: 0x80131505)
   at UIAutomationClient.IUIAutomationElement.GetCurrentPropertyValueEx(Int32 propertyId, Int32 ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property, Boolean ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property)
   at ExcelDna.IntelliSense.PopupListWatcher.UpdateSelectedItem(AutomationElement newSelectedItem)
    DateTime=2016-12-07T14:43:08.9494852Z
ExcelDna.IntelliSense Verbose: 4 : PopupList PopupListElementSelectedHandler on thread 1
    DateTime=2016-12-07T14:43:08.9837960Z
ExcelDna.IntelliSense Verbose: 4 : PopupList SelectedItemChanged SUM ListBounds: 504,262,114,189~~~~

**Comment by govert** [comment]
Thank you.
At least that gives us an error code - but the most relevant link seemed like this one: https://social.msdn.microsoft.com/Forums/en-US/787d65c6-61c4-496a-a1be-a742b3200a14/global-hook-uiautomation-cannot-handle-wpf-controls-correctly-exception-operation-timed-out?forum=windowsaccessibilityandautomation

Have you got any WPF stuff loaded into Excel?
(The IntelliSense display uses WinForms, not WPF.)

**Comment by mfriedy** [comment]
No, no other addins excel except ExcelDNA. The only other piece of information I can provide is that when excel hangs, the white/blank formula autocomplete box persists across any app you switch into.

**Comment by govert** [comment]
Just to bookmark some links about UI Automation timeouts:
* http://stackoverflow.com/questions/31135839/is-there-any-internal-timeout-in-microsoft-uiautomation
* https://msdn.microsoft.com/en-us/library/windows/desktop/hh437206(v=vs.85).aspx

I'm wondering why the call would block Excel's thread...
(It's worth checking again that we really are running on the intended UI automation thread, and not the Excel main thread, at this point.)

**Comment by jimmcslim** [comment]
I have also noticed this starting recently. Not sure if it is related to an Office update, but for the record I am using:

* Office Home and Business 2016 64-bit
* Version 1611 (Build 7571.2075)
* About dialog says the version is 16.0.7571.7063

I'll check some user's laptops to see if they are using the same version - they possibly are using older versions that aren't autoupdating - hopefully it is just my own workstation that has the issue at the moment as this makes the Intellisense add-in unusable unfortunately (which is a great shame as it makes our custom functions so much more usable and discoverable!)

Our add-in does have a custom task pane that uses WPF, for what it is worth.

**Comment by govert** [comment]
@jimmcslim Thank you for the updated report.
So far it seems this problem is only with up-to-date Excel 2016 64-bit.

I've not yet had a chance to try to replicate yet - I still need to set up an Excel 2016 64-bit VM somewhere.

@mfriedy Could you perhaps confirm your exact Excel and Windows versions?

**Comment by mfriedy** [comment]
@govert  i am on the same setup as @jimmcslim. My Excel is 16.0.7571.7063 64-bit too, and Windows Version 1607 (build 14393.576). 

It may be helpful to say that our add-in only has non-macro functions, so none of our own display UI features (e.g. WPF code). 

Later this week I have 24hrs of flying - I have the project compiling and running from source on my machine so am more than happy to do some digging around if it will help? Is there a particular area of the code you think is probably to blame?

**Comment by govert** [comment]
@mfriedy Thanks for confirming the versions.

>  Is there a particular area of the code you think is probably to blame?

I suspect the problematic code is inside Excel. I don't know if we can work around it yet.

**Comment by govert** [comment]
@mfriedy One step might be to experiment with the UI Automation `TransactionTimeout` exposed on `IUIAutomation2`. (How to get to this interface etc....? I don't know.) But the default value of 20s sounds close to what you're reporting.

**Comment by govert** [comment]
@mfriedy Maybe an even simpler step would be to wrap a StopWatch around the call that fails, and jsut log and understand the timing of the call (and whether it generally succeeds) in different settings - both on a 'problematic' Excel and something like Excel 2013 32-bit that seems to work reliably. I.e. we add a bit more instrumentation around the problem code. 

I guess some weird things could be going on with the popup list, that causes our bounds query to go wrong. E.g. Excel could have recycled or swapped out the window for some reason, and now the one we think was the popup Window no longer exists. Many, many things could be going on behind the scenes. 

What concerns me most is that Excel itself seems to be stuck trying to process our UI Automation request. Is there some internal race condition?

Finally, one might try to trigger the same issue with a standard UI Automation tool like White (https://github.com/TestStack/White). That would make it easier to report to Microsoft as a bug. But it might also not reproduce what we see, because Excel-DNA is running in-process.

So I'm not stuck for ideas on how to proceed, but everything related to the UI Automation and monitoring the Excel UI is incredibly frustrating and tedious to work on - that's why the IntelliSense extension took years to get to this point.

**Comment by mfriedy** [comment]
@govert - thanks, i'll see what I may be able to stumble upon. I actually met some of the excel development team a couple of weeks ago - I am trying to get back in touch and (if successful) will see if they can pinpoint the change that caused this to stop working.

**Comment by jimmcslim** [comment]
To contribute some more info to the investigation; I've checked with some of my users who don't have as frequent Office updates in place, and who are not experiencing the issue.

They are on Office 2016 16.0.7426.1051 (64-bit, Home & Business).

**Comment by govert** [comment]
I've now had this reported on a 32-bit Excel 2016 too.
The delays started when updated to version 16.0.7571.7063.
Previously there were no problems.

**Comment by peterludvik** [comment]
I had also noticed the delays in the DNA.Intellisence, running
Windows 10 Pro, Excel 2016 version 1611(Build 7571.2075)
Otherwise fantastically useful tool - still exploring the possibilities!

**Comment by andrewkittredge** [comment]
I just started having this problem.  Excel 32-bit.  Version 1611, Build 7571.2075

**Comment by andrewkittredge** [comment]
See this issue related to visual studio, it might be the same thing.  MSFT claims it is fixed.

https://developercommunity.visualstudio.com/content/problem/4760/intellisense-frequently-hangs.html

**Comment by asndev** [comment]
We are running into the same problem, looks like it occurred via the last Excel update.
(W10, 32bit, Version 1611, Build 7571.2075)

Workaround with the environment variable is working great so far :-)
Lets see what the next Excel update will reveal.

**Comment by opus131** [comment]
I see the same issue, with the same build (W10, 32bit, Version 1611, Build 7571.2075). 

@govert is my undestand correct that this is not something that can be resolved from inside the add-in - and that we'll just have to hope that a future Excel release will fix this?

**Comment by govert** [comment]
@opus131 I have not had a chance to look at this more closely. I don't have an Excel 2016 environment set up at the moment.

My understanding is that Excel 'hangs' during a call we make to UI Automation from another thread (which should always be safe). So I'm nearly sure the problem code is inside Excel, but maybe there is some workaround we could implement - the UI Automation timeout might be one try. However, working on and debugging any of the UI Automation stuff is incredibly frustrating and time consuming. 

I've had reports of various bugs appearing and later being fixed with these rolling Excel 2016 updates, so there is definitely a QA problem inside the Excel team. I'll probably wait for one or two update cycles of Excel to see if Microsoft does anything about it, before spending more of my time on it.

If anyone has a channel to Microsoft to report or escalate this issue, please use it!  One large bank raising the issue with their sales contact might sort out the issue very quickly.

**Comment by govert** [comment]
@jimmcslim  @peterludvik @andrewkittredge @asndev @opus131 
Could I perhaps ask that you configure the verbose logging, and post the relevant exception bit from the log after running into the delay? I might have a plan for a workaround ...

To add verbose logging, create a file called ExcelDna.IntelliSense.xll.config next to your ExcelDna.IntelliSense.xll.
(For 64-bit Excel you might have ExcelDna.IntelliSense64.xll and would call it ExcelDna.IntelliSense64.xll.config. For an integrated configuration, this would go in your .xll.config file.)

Put this into the config file:

```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <system.diagnostics>
    <trace autoflush="true" indentsize="4"/>
    <sources>
      <source name="ExcelDna.IntelliSense" switchValue="Verbose">
        <listeners>
          <remove name="Default" />
          <add name="File" type="System.Diagnostics.TextWriterTraceListener" initializeData="ExcelDna.IntelliSense.log" traceOutputOptions = "DateTime" >
            <filter type="System.Diagnostics.EventTypeFilter" initializeData="All"/>
          </add>
        </listeners>
      </source>
    </sources>
  </system.diagnostics>
</configuration>
```

I'm looking for the region around a line that looks like this:

`ExcelDna.IntelliSense Warning: 4 : PopupList - Could not update selected item: `

**Comment by opus131** [comment]
Thank you very much for your response @govert.

Unfortunately I cannot see a warning in the log, I'll upload the whole output here (what I did: opened Excel, triggered the intellisense, see it hang, closed Excel again).

I hope this contain any useful information for you.

[ExcelDna.IntelliSense.zip](https://github.com/Excel-DNA/IntelliSense/files/711113/ExcelDna.IntelliSense.zip)

**Comment by asndev** [comment]
I did the same and didn't see any warning either

FYI: [intellisense.zip](https://github.com/Excel-DNA/IntelliSense/files/711133/intellisense.zip)

**Comment by opus131** [comment]
Update: played around with it some more, and this time I saw two warnings:


```
ExcelDna.IntelliSense Verbose: 2 : >> (FunctionListSelectedItemChange): [ExcelDna.IntelliSense.UIState+FunctionList] -> [ExcelDna.IntelliSense.UIState+FunctionList]
    DateTime=2017-01-17T13:24:57.3824785Z
ExcelDna.IntelliSense Verbose: 2 : !> ExcelToolTip ToolTipChanged Received: Hide:0x2625842 with state: FunctionList
    DateTime=2017-01-17T13:24:57.3824785Z
ExcelDna.IntelliSense Verbose: 2 : >> (FormulaEditExcelToolTipChange): [ExcelDna.IntelliSense.UIState+FunctionList] -> [ExcelDna.IntelliSense.UIState+FunctionList]
    DateTime=2017-01-17T13:24:57.3824785Z
ExcelDna.IntelliSense Verbose: 4 : PopupList PopupListElementSelectedHandler on thread 1
    DateTime=2017-01-17T13:24:57.3864797Z
ExcelDna.IntelliSense Warning: 4 : PopupList - Could not update selected item: System.Runtime.InteropServices.COMException (0x80131505): Operation timed out. (Exception from HRESULT: 0x80131505)
   at UIAutomationClient.IUIAutomationElement.GetCurrentPropertyValueEx(Int32 propertyId, Int32 ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property, Boolean ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property)
   at ExcelDna.IntelliSense.PopupListWatcher.UpdateSelectedItem(AutomationElement newSelectedItem)
    DateTime=2017-01-17T13:25:17.3904699Z
ExcelDna.IntelliSense Verbose: 4 : FormulaEdit - Text changed in InCellEdit
    DateTime=2017-01-17T13:25:17.3904699Z
ExcelDna.IntelliSense Verbose: 4 : >>>> FormulaEditWatcher.UpdateFormula on thread 1
    DateTime=2017-01-17T13:25:17.3904699Z
ExcelDna.IntelliSense Verbose: 4 : >>>> FormulaEditWatcher.UpdateFormula CurrentPrefix: =GetSt
    DateTime=2017-01-17T13:25:17.3904699Z
```


And this one:


```
ExcelDna.IntelliSense Verbose: 4 : FormulaEdit UpdateEditState - Focus: InCellEdit
    DateTime=2017-01-17T13:28:36.5090979Z
ExcelDna.IntelliSense Verbose: 4 : FormulaEdit - Text changed in InCellEdit
    DateTime=2017-01-17T13:28:38.5286305Z
ExcelDna.IntelliSense Verbose: 4 : >>>> FormulaEditWatcher.UpdateFormula on thread 1
    DateTime=2017-01-17T13:28:38.5286305Z
ExcelDna.IntelliSense Verbose: 4 : >>>> FormulaEditWatcher.UpdateFormula CurrentPrefix: =G
    DateTime=2017-01-17T13:28:38.5286305Z
ExcelDna.IntelliSense Warning: 2 : SingleThreadSynchronizationContext ### Unhandled exception (Thread 11) - System.Windows.Automation.ElementNotAvailableException: Element not available ---> System.Runtime.InteropServices.COMException: An event was unable to invoke any of the subscribers (Exception from HRESULT: 0x80040201)
   at UIAutomationClient.IUIAutomationElement.GetCurrentPropertyValueEx(Int32 propertyId, Int32 ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property, Boolean ignoreDefaultValue)
   --- End of inner exception stack trace ---
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property, Boolean ignoreDefaultValue)
   at System.Windows.Automation.AutomationElement.GetCurrentPropertyValue(AutomationProperty property)
   at ExcelDna.IntelliSense.FormulaEditWatcher.<UpdateEditState>b__42_0(Object moveOnlyObj)
   at ExcelDna.IntelliSense.SingleThreadSynchronizationContext.RunOnCurrentThread()
    DateTime=2017-01-17T13:28:38.5286305Z
ExcelDna.IntelliSense Verbose: 2 : !> ExcelToolTip ToolTipChanged Received: Show:0x4067774 with state: FormulaEdit
    DateTime=2017-01-17T13:28:38.5286305Z
ExcelDna.IntelliSense Verbose: 4 : PopupList window show
    DateTime=2017-01-17T13:28:38.5286305Z
```

**Comment by govert** [comment]
@opus131 That first one looks similar to what I expect, and has the delay in. I'll try to make a version to try a workaround in the next few days.

**Comment by govert** [comment]
@asndev @opus131 I've made a new release (v1.0.1) that might be able to workaround this issue. I've just added an extra check for a UI Automation event that Excel might (unexpectedly) be raising on the main thread.

Please have a look, and if the delays still appear, perhaps you can make a new log.

**Comment by jimmcslim** [comment]
I have tried the new code (I'm using the embedded .dll version, so have temporarily removed the Nuget package and referencing my compiled 1.0.1 assembly directly).

I'm still getting hangs unfortunately (Excel 2016, version 1611 build 7571.2109 according to the Account page, or 16.0.7571.7095 according to About Excel). I don't have a log to contribute, however if I break into the debugger I am seeing that it is waiting at either line 404 or 405 of FormulaEditWatcher.cs

Not sure I can enable logging easily, but I'll try again later today - I may even have a futz around with the code to see if I can contribute a useful fix.

**Comment by opus131** [comment]
Thank you very much for the quick update! Unfortunately the issue persists. 

New log attached:

[ExcelDna.IntelliSense.zip](https://github.com/Excel-DNA/IntelliSense/files/713270/ExcelDna.IntelliSense.zip)

**Comment by govert** [comment]
@opus131 Thank you - I'm not seeing the long delay in your log though. (That doesn't mean it wasn't there, but in the previous logs I could see the 20 s delay followed by a UI Automation Timeout, which is not there anymore.)

**Comment by asndev** [comment]
Latest logfile with IntelliSense 1.0.1.

[ExcelDna.IntelliSense.log.zip](https://github.com/Excel-DNA/IntelliSense/files/713604/ExcelDna.IntelliSense.log.zip)

Still can't find anything interesting yet. Tried to comment the lines  (and many more) in the FormulaEditWatcher mentioned by @jimmcslim, but it still freezes for me.

**Comment by govert** [comment]
@asndev Thank you - I also don't see any long delay visible in that log file...

**Comment by asndev** [comment]
FYI
In FormulaEditWatcher constructor, commenting this line `            _windowWatcher.InCellEditWindowChanged += _windowWatcher_InCellEditWindowChanged;
` has the effect, that it doesn't freeze anymore.

**Comment by junos** [comment]
@govert When will the fix release at NuGet? thanks!

**Comment by govert** [comment]
@junos I'm afraid the attempted fix didn't really work :-(

Perhaps a new Excel 2016 update will improve things....

**Comment by govert** [comment]
I had not seen this before, but the list of updates for the November 2016 update of Excel 2016 confirms that there were changes made to the accessibility support (which the Excel-DNA IntelliSense uses to track the formula state). This link has some more details on the changes - https://support.office.com/en-us/article/What-s-new-in-accessibility-for-Excel-51fcb17a-b15b-4b13-ae04-d4f38ece3f78?ui=en-US&rs=en-US&ad=US

Nothing on the list jumps out at me, but they did some work on this area and it seems our IntelliSense extension suffered collateral damage.

**Comment by govert** [comment]
I've made another new release, hopefully giving a proper fix to the Excel 2016 issues.
The new release is available on GitHub here: https://github.com/Excel-DNA/IntelliSense/releases/tag/v1.0.2
and (@junos) as an updated NuGet package (ExcelDna.IntelliSense version 1.0.2).

The internals have changed significantly, since I've removed the UI Automation dependency.
Please let me know whether this fixes the Excel 2016 delays, and whether I've introduced any regressions.

**Comment by jimmcslim** [comment]
Hi @govert I have given this a quick go, and looks good from my perspective!

Running Excel version 1609, build 7369.2130, this is the latest release on the Office 365 ProPlus Deferred channel as of May 9, 2017.

Thanks for your work on this, my users will definitely appreciate the intellisense support coming back!

**Comment by andrewkittredge** [comment]
It works on my machine.  We are testing it internally on our sundry versions of Excel.

**Comment by amibar** [comment]
@govert, I tested it on Excel 2016 as part of Offce 365 Business and it doesn't hang anymore which is great.

However, I just noticed that when I type a formula the IntelliSense is updated 1 char too late.

Let's say I am typing the following chars. each step contains the text I already typed.

1. =AddThem
2. =AddThem(
3. =AddThem(1
4. =AddThem(1,
5. =AddThem(1,2

After step 2 the tool tip doesn't appear at all, it appear after step 3
After step 4 the second arguments tools tip doesn't appear, it appear after step 5

It look like the tools tip looks at the previous text of the formula and not the current one.

I don't know if this issue belongs here so you can fix it all together, or it should be another issue, your choice.

Anyway, great work with all this Excel AddIns, it helps me a lot.

**Comment by govert** [comment]
@amibar Could you open a new issue for this?
